x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/ClientDataFrameIndexer.java

```diff
@@ -0,0 +1,591 @@
+                            updateConfigListener.onResponse(null);
+                        },
+                        failure -> {
+                            String msg = DataFrameMessages.getMessage(
+                                DataFrameMessages.FAILED_TO_RELOAD_TRANSFORM_CONFIGURATION,
+                                getJobId());
+                            logger.error(msg, failure);
+                            // If the transform config index or the transform config is gone, something serious occurred
+                            // We are in an unknown state and should fail out
+                            if (failure instanceof ResourceNotFoundException) {
+                                updateConfigListener.onFailure(new TransformConfigReloadingException(msg, failure));
+                            } else {
+                                auditor.warning(getJobId(), msg);
+                                updateConfigListener.onResponse(null);
+                            }
+                        }
+                    ));
+                } else {
+                    updateConfigListener.onResponse(null);
+                }
+            },
+            listener::onFailure
+        );
+
+        // If we are not on the initial batch checkpoint and its the first pass of whatever continuous checkpoint we are on,
+        // we should verify if there are local changes based on the sync config. If not, do not proceed further and exit.
+        if (transformTask.getCheckpoint() > 0 && initialRun()) {
+            sourceHasChanged(ActionListener.wrap(
+                hasChanged -> {
+                    hasSourceChanged = hasChanged;
+                    if (hasChanged) {
+                        changesLastDetectedAt = Instant.now();
+                        logger.debug("[{}] source has changed, triggering new indexer run.", getJobId());
+                        changedSourceListener.onResponse(null);
+                    } else {
+                        logger.trace("[{}] source has not changed, finish indexer early.", getJobId());
+                        // No changes, stop executing
+                        listener.onResponse(false);
+                    }
+                },
+                failure -> {
+                    // If we failed determining if the source changed, it's safer to assume there were changes.
+                    // We should allow the failure path to complete as normal
+                    hasSourceChanged = true;
+                    listener.onFailure(failure);
+                }
+            ));
+        } else {
+            hasSourceChanged = true;
+            changedSourceListener.onResponse(null);
+        }
+    }
+
+    public CheckpointProvider getCheckpointProvider() {
+        return checkpointProvider;
+    }
+
+    public Instant getChangesLastDetectedAt() {
```

Comment:
```suggestion
    Instant getChangesLastDetectedAt() {
```
Can this be package private?